packages/kbn-subscription-tracking/src/subscription_elements.tsx

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import React from 'react';
+import { EuiLink, EuiButton, EuiButtonEmpty } from '@elastic/eui';
+import type { EuiLinkProps, EuiButtonEmptyProps, EuiButtonProps } from '@elastic/eui';
+import { useGoToSubscription } from './use_go_to_subscription';
+import { useImpression } from './use_impression';
+import type { SubscriptionContextData } from '../types';
+
+interface CommonProps {
+  /** The context information for this subscription element */
+  subscriptionContext: SubscriptionContextData;
+}
+
+export type SubscriptionLinkProps = EuiLinkProps & CommonProps;
+
+/**
+ * Wrapper around `EuiLink` that provides subscription events
+ */
+export function SubscriptionLink({
```

Comment:
Basically `SubscriptionLink`, `SubscriptionButton` and `SubscriptionButtonEmpty` have exactly the same body with the only difference in the underlying component used. As a generalisation it could be one component like `Subscription` with a `component` prop which could be set to `EuiLink`, `EuiButton` or `EuiButtonEmpty` or event something else with the same interface. TS generics could help here.

Reply:
Good idea @maximpn ! I tried this before but could not get the prop types to align nicely since some of them are mutually exclusive. 

The current approach also allows for tweaking of individual behaviours in case it's necessary in the future. Also, all repetition is contained in this one file.

Wdyt?

Reply:
I'm fine with per component approach but just highlighted it breaks DRY pricnciple. It looks simple as properly type `restProps` though it can be tricky with TS generics.